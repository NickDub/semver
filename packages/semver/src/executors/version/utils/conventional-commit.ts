import * as conventionalChangelog from 'conventional-changelog';
import { WriteChangelogConfig } from '../schema';

export function createConventionalCommitStream(
  config: WriteChangelogConfig,
  newVersion: string,
) {
  return conventionalChangelog(
    {
      ...(typeof config.preset === 'string' ? { preset: config.preset } : {}),
      ...(typeof config.preset === 'object' ? { config: config.preset } : {}),
      tagPrefix: config.tagPrefix,
      pkg: {
        path: config.projectRoot,
      },
    },
    { version: newVersion },
    /// @ts-expect-error - Partially typed API
    { path: config.projectRoot },
  );
}
