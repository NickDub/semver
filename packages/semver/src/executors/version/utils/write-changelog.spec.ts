import * as fs from 'fs';
import * as Stream from 'stream';

import { createConventionalCommitStream } from './conventional-commit';
import writeChangelog from './write-changelog';
import type { WriteChangelogConfig } from '../schema';

jest.mock('./conventional-commit');

const mockCreateConventionalCommitStream =
  createConventionalCommitStream as jest.MockedFunction<
    typeof createConventionalCommitStream
  >;

const config: WriteChangelogConfig = {
  changelogHeader: '# Changelog',
  projectRoot: './',
  preset: 'angular',
  dryRun: false,
  changelogPath: 'CHANGELOG.md',
  tagPrefix: 'button',
};

describe(writeChangelog, () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation();
    jest.spyOn(console, 'info').mockImplementation();
    jest.spyOn(fs, 'writeFileSync').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('handle errors', () => {
    beforeEach(async () => {
      mockCreateConventionalCommitStream.mockReturnValue(
        new Stream.Readable({
          read() {
            this.emit('error', 'ðŸ’¥');
          },
        }),
      );
      await writeChangelog(config, '0.0.1');
    });

    it('should print a console.warn', async () => {
      expect(console.warn).toHaveBeenCalledWith(
        'changelog creation failed',
        'ðŸ’¥',
      );
    });

    it('should not write a changelog file', async () => {
      expect(fs.writeFileSync).not.toHaveBeenCalled();
    });
  });

  describe('--dryRun', () => {
    const version = '1.0.0';

    beforeEach(async () => {
      mockCreateConventionalCommitStream.mockImplementation(
        jest.requireActual('./conventional-commit')
          .createConventionalCommitStream,
      );

      await writeChangelog({ ...config, dryRun: true }, version);
    });

    it('should not write a changelog file', async () => {
      expect(fs.writeFileSync).not.toHaveBeenCalled();
    });

    it('should print a console.info with the changelog contents without the header', async () => {
      expect(console.info).toHaveBeenCalledWith(
        expect.stringContaining(`## ${version}`),
      );
      expect(console.info).toHaveBeenCalledWith(
        expect.not.stringContaining(config.changelogHeader),
      );
    });
  });

  describe('--preset', () => {
    const version = '1.0.0';

    beforeEach(async () => {
      mockCreateConventionalCommitStream.mockImplementation(
        jest.requireActual('./conventional-commit')
          .createConventionalCommitStream,
      );

      await writeChangelog(
        {
          ...config,
          preset: {
            types: [
              { type: 'feat', section: 'Awesome Features' },
              { type: 'fix', section: 'Important Fixes' },
              { type: 'chore', hidden: true },
              { type: 'docs', hidden: true },
              { type: 'style', hidden: true },
              { type: 'refactor', hidden: true },
              { type: 'perf', hidden: true },
              { type: 'test', hidden: true },
            ],
          },
          dryRun: true,
        },
        version,
      );
    });

    it('should load custom preset', () => {
      expect(
        mockCreateConventionalCommitStream.mock.calls[0][0].preset,
      ).toMatchSnapshot();
    });

    it('should print changelog', () => {
      expect((console.info as jest.Mock).mock.calls[0][0]).toContain(
        'Awesome Features',
      );
      expect((console.info as jest.Mock).mock.calls[0][0]).toContain(
        'Important Fixes',
      );
    });
  });
});
