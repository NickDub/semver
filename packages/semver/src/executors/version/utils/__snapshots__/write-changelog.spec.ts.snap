// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeChangelog --preset should load custom preset 1`] = `
Object {
  "conventionalChangelog": Object {
    "parserOpts": Object {
      "breakingHeaderPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!: \\(\\.\\*\\)\\$/,
      "headerCorrespondence": Array [
        "type",
        "scope",
        "subject",
      ],
      "headerPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!\\?: \\(\\.\\*\\)\\$/,
      "issuePrefixes": Array [
        "#",
      ],
      "noteKeywords": Array [
        "BREAKING CHANGE",
        "BREAKING-CHANGE",
      ],
      "revertCorrespondence": Array [
        "header",
        "hash",
      ],
      "revertPattern": /\\^\\(\\?:Revert\\|revert:\\)\\\\s"\\?\\(\\[\\\\s\\\\S\\]\\+\\?\\)"\\?\\\\s\\*This reverts commit \\(\\\\w\\*\\)\\\\\\./i,
    },
    "writerOpts": Object {
      "commitGroupsSort": [Function],
      "commitPartial": "*{{#if scope}} **{{scope}}:**
{{~/if}} {{#if subject}}
  {{~subject}}
{{~else}}
  {{~header}}
{{~/if}}

{{~!-- commit link --}}{{~#if hash}} {{#if @root.linkReferences~}}
  ([{{shortHash}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/commit/{{hash}}))
{{~else}}
  {{~shortHash}}
{{~/if}}{{~/if}}

{{~!-- commit references --}}
{{~#if references~}}
  , closes
  {{~#each references}} {{#if @root.linkReferences~}}
    [
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}{{this.prefix}}{{this.issue}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/issues/{{this.issue}})
  {{~else}}
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}{{this.prefix}}{{this.issue}}
  {{~/if}}{{/each}}
{{~/if}}

",
      "commitsSort": Array [
        "scope",
        "subject",
      ],
      "footerPartial": "",
      "groupBy": "type",
      "headerPartial": "## {{#if @root.linkCompare~}}
  [{{version}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/compare/{{previousTag}}...{{currentTag}})
{{~else}}
  {{~version}}
{{~/if}}
{{~#if title}} \\"{{title}}\\"
{{~/if}}
{{~#if date}} ({{date}})
{{/if}}
",
      "mainTemplate": "{{> header}}

{{#if noteGroups}}
{{#each noteGroups}}

### ⚠ {{title}}

{{#each notes}}
* {{#if commit.scope}}**{{commit.scope}}:** {{/if}}{{text}}
{{/each}}
{{/each}}
{{/if}}
{{#each commitGroups}}

{{#if title}}
### {{title}}

{{/if}}
{{#each commits}}
{{> commit root=@root}}
{{/each}}

{{/each}}
",
      "noteGroupsSort": "title",
      "notesSort": [Function],
      "transform": [Function],
    },
  },
  "gitRawCommitsOpts": Object {
    "ignore": undefined,
    "noMerges": null,
  },
  "parserOpts": Object {
    "breakingHeaderPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!: \\(\\.\\*\\)\\$/,
    "headerCorrespondence": Array [
      "type",
      "scope",
      "subject",
    ],
    "headerPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!\\?: \\(\\.\\*\\)\\$/,
    "issuePrefixes": Array [
      "#",
    ],
    "noteKeywords": Array [
      "BREAKING CHANGE",
      "BREAKING-CHANGE",
    ],
    "revertCorrespondence": Array [
      "header",
      "hash",
    ],
    "revertPattern": /\\^\\(\\?:Revert\\|revert:\\)\\\\s"\\?\\(\\[\\\\s\\\\S\\]\\+\\?\\)"\\?\\\\s\\*This reverts commit \\(\\\\w\\*\\)\\\\\\./i,
  },
  "recommendedBumpOpts": Object {
    "parserOpts": Object {
      "breakingHeaderPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!: \\(\\.\\*\\)\\$/,
      "headerCorrespondence": Array [
        "type",
        "scope",
        "subject",
      ],
      "headerPattern": /\\^\\(\\\\w\\*\\)\\(\\?:\\\\\\(\\(\\.\\*\\)\\\\\\)\\)\\?!\\?: \\(\\.\\*\\)\\$/,
      "issuePrefixes": Array [
        "#",
      ],
      "noteKeywords": Array [
        "BREAKING CHANGE",
        "BREAKING-CHANGE",
      ],
      "revertCorrespondence": Array [
        "header",
        "hash",
      ],
      "revertPattern": /\\^\\(\\?:Revert\\|revert:\\)\\\\s"\\?\\(\\[\\\\s\\\\S\\]\\+\\?\\)"\\?\\\\s\\*This reverts commit \\(\\\\w\\*\\)\\\\\\./i,
    },
    "whatBump": [Function],
  },
  "writerOpts": Object {
    "commitGroupsSort": [Function],
    "commitPartial": "*{{#if scope}} **{{scope}}:**
{{~/if}} {{#if subject}}
  {{~subject}}
{{~else}}
  {{~header}}
{{~/if}}

{{~!-- commit link --}}{{~#if hash}} {{#if @root.linkReferences~}}
  ([{{shortHash}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/commit/{{hash}}))
{{~else}}
  {{~shortHash}}
{{~/if}}{{~/if}}

{{~!-- commit references --}}
{{~#if references~}}
  , closes
  {{~#each references}} {{#if @root.linkReferences~}}
    [
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}{{this.prefix}}{{this.issue}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/issues/{{this.issue}})
  {{~else}}
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}{{this.prefix}}{{this.issue}}
  {{~/if}}{{/each}}
{{~/if}}

",
    "commitsSort": Array [
      "scope",
      "subject",
    ],
    "footerPartial": "",
    "groupBy": "type",
    "headerPartial": "## {{#if @root.linkCompare~}}
  [{{version}}]({{~@root.host}}/{{#if this.owner}}{{~this.owner}}{{else}}{{~@root.owner}}{{/if}}/{{#if this.repository}}{{~this.repository}}{{else}}{{~@root.repository}}{{/if}}/compare/{{previousTag}}...{{currentTag}})
{{~else}}
  {{~version}}
{{~/if}}
{{~#if title}} \\"{{title}}\\"
{{~/if}}
{{~#if date}} ({{date}})
{{/if}}
",
    "mainTemplate": "{{> header}}

{{#if noteGroups}}
{{#each noteGroups}}

### ⚠ {{title}}

{{#each notes}}
* {{#if commit.scope}}**{{commit.scope}}:** {{/if}}{{text}}
{{/each}}
{{/each}}
{{/if}}
{{#each commitGroups}}

{{#if title}}
### {{title}}

{{/if}}
{{#each commits}}
{{> commit root=@root}}
{{/each}}

{{/each}}
",
    "noteGroupsSort": "title",
    "notesSort": [Function],
    "transform": [Function],
  },
}
`;
